homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,263 @@
+from homeassistant.util.dt import utcnow
+
+from . import WhirlpoolData
+from .const import DOMAIN
+
+TANK_FILL = {
+    "0": "Unknown",
+    "1": "Empty",
+    "2": "25%",
+    "3": "50%",
+    "4": "100%",
+    "5": "Active",
+}
+
+MACHINE_STATE = {
+    MachineState.Standby: "Standby",
+    MachineState.Setting: "Setting",
+    MachineState.DelayCountdownMode: "Delay Countdown",
+    MachineState.DelayPause: "Delay Paused",
+    MachineState.SmartDelay: "Smart Delay",
+    MachineState.SmartGridPause: "Smart Grid Pause",
+    MachineState.Pause: "Pause",
+    MachineState.RunningMainCycle: "Running Maincycle",
+    MachineState.RunningPostCycle: "Running Postcycle",
+    MachineState.Exceptions: "Exception",
+    MachineState.Complete: "Complete",
+    MachineState.PowerFailure: "Power Failure",
+    MachineState.ServiceDiagnostic: "Service Diagnostic Mode",
+    MachineState.FactoryDiagnostic: "Factory Diagnostic Mode",
+    MachineState.LifeTest: "Life Test",
+    MachineState.CustomerFocusMode: "Customer Focus Mode",
+    MachineState.DemoMode: "Demo Mode",
+    MachineState.HardStopOrError: "Hard Stop or Error",
+    MachineState.SystemInit: "System Initialize",
+}
+
+CYCLE_FUNC = [
+    (WasherDryer.get_cycle_status_filling, "Cycle Filling"),
+    (WasherDryer.get_cycle_status_rinsing, "Cycle Rinsing"),
+    (WasherDryer.get_cycle_status_sensing, "Cycle Sensing"),
+    (WasherDryer.get_cycle_status_soaking, "Cycle Soaking"),
+    (WasherDryer.get_cycle_status_spinning, "Cycle Spinning"),
+    (WasherDryer.get_cycle_status_washing, "Cycle Washing"),
+]
+
+
+ICON_D = "mdi:tumble-dryer"
+ICON_W = "mdi:washing-machine"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def washer_state(washer: WasherDryerClass) -> str | None:
+    """Determine correct states for a washer."""
+
+    if washer.washdry.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        return "Door open"
+
+    machine_state = washer.washdry.get_machine_state()
+    machine_cycle = None
+
+    for func, cycle_name in CYCLE_FUNC:
+        if func(washer.washdry):
+            machine_cycle = cycle_name
+            break
+
+    if machine_state == MachineState.RunningMainCycle and machine_cycle:
+        return machine_cycle
+
+    return MACHINE_STATE.get(machine_state, STATE_UNKNOWN)
+
+
+def completion_time(washer: WasherDryerClass) -> datetime | None:
+    """Calculate the time stamp for completion."""
+    machine_state = washer.washdry.get_machine_state()
+    if (
+        machine_state.value in {MachineState.Complete.value, MachineState.Standby.value}
+        and washer.running
+    ):
+        washer.running = False
+        washer.timestamp = utcnow()
+
+    if machine_state is MachineState.RunningMainCycle:
+        washer.running = True
+        washer.timestamp = utcnow() + timedelta(
+            seconds=int(
+                washer.washdry.get_attribute("Cavity_TimeStatusEstTimeRemaining")
+            )
+        )
+    return washer.timestamp
```

Comment:
https://github.com/home-assistant/core/pull/84851#discussion_r1060221206

@allenporter - picking up where we left off:
If I do it this way, the entity shows "unknown" when the washer is not running.

The way it is implemented, it counts down until the cycle ends, then returns the timestamp for when they cycle ended (not None) from the stored timestamp.
It then shows how long "until it ends", then when it ends, it shows "how long since it ended".

I originally had it like your example with return utcnow() instead of return None, but that ends up updating the end time when the API receives a state change message - I needed to store utcnow() at the state change.

Am I way off base here?

Reply:
Ah i was reviewing this in parallel and missed your comment... copying my comment here:

Thanks for the explanation of this sensor and no you aren't way off, i just have two things I see a bit differently:

(1) Make it a separate entity so that the other sensor types don't also have these fields (which aren't used). I think it would make sense not to use the value fn for that entity and just keep the state within that sensor

(2) What do you want the value to be on home assistant startup? My impression is you can't set one, which is why my naive example punts on trying to get fancier and storing a previously estimated completion time -- I'm ok with only having a value when its not active, because I don't think it matters when the state is completed or not running. We don't need the estimated completion time anymore since we already know it has passed.

Reply:
#2, it when the sensor initializes "washer.running = True", this way if the machine is NOT running, we return utcnow() and set washer.running = False.  If it's actually running, we get the estimated end time.

#1, I suppose it's a toss up on how we do this.  We run "special" code either way.  Adding a separate entity would tighten up the access to the variables.  Probably some python PEP out there saying which way to go.... I'm just not familiar enough with them.

I'll do whichever way is preferred :)

Reply:
(1) We should not use a value function that has side effects. That is not expected.
(2) My take is this seems like an arbitrary value, and we know it isn't correct. Therefore, setting `None` seems more clear to indicate that and i'd feel comfortable merging that.  If you want to go now for unknown then we'll need someone else to review and  I can tag with `Second opinion needed` to flag for another reviewer. You can also start with None, and then send another PR to add the behavior you are proposing if you want to keep moving forward incrementally.

Reply:
Ok - I've made a separate Entity for the time function and defaulted to None for self._timestamp.
Is this what you had in mind?